provider/pkg/provider/remote/copy.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/archive"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/asset"
+)
+
+type Copy struct{}
+
+var _ = (infer.Annotated)((*Copy)(nil))
+
+// Copy implements Annotate which allows you to attach descriptions to the Copy resource.
+func (c *Copy) Annotate(a infer.Annotator) {
+	a.Describe(&c, "Copy an Asset or Archive to a remote host.")
+}
+
+type CopyInputs struct {
+	Connection *Connection      `pulumi:"connection" provider:"secret"`
+	Triggers   *[]interface{}   `pulumi:"triggers,optional" providers:"replaceOnDelete"`
+	Asset      *asset.Asset     `pulumi:"asset,optional"`
+	Archive    *archive.Archive `pulumi:"archive,optional"`
```

Comment:
Oooh interesting! I just looked at the definitions of `Asset` and `Archive` and saw that they're separate.

Reply:
Yeah, the design of Assets overall is pretty confusing on both the provider and consumer sides 🤷

Reply:
Here's an example from azure-native:

https://github.com/pulumi/pulumi-azure-native/blob/f607b7a1e27913aef9672b75913bdf341c46c118/provider/pkg/resources/customresources/custom_storage.go#L659-L678

Reply:
This works a bit differently because it's using raw property values rather than the new framework. Perhaps this is a missing feature in the go-provider to detect which kind of asset we've got at runtime?

Reply:
This discussion led to filing https://github.com/pulumi/pulumi-go-provider/issues/237